EMP-406 create prepare disk test script

raysiudak opened the pull request.

raysiudak committed:
Add initial disk prepare script

- This includes the basic components needed to execute
  prepare_disk script tests.
- Add test script execution to build_for_ultimaker
- Add test script execution to CI

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add prepare disk tests currently implemented on master

- test_execute_prepare_disk_sha512_nok
- test_execute_prepare_disk_grow_rootfs_ok
- test_execute_prepare_disk_resize_not_needed_ok

Leave them in the toolbox_image test, cleaning that up will be done later.

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add disk_prepare test: grow boot partition

Contributes to EMP-389, implements EMP-417

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add disk_prepare test: grow boot overlapping rootfs fails

Contributes to EMP-389, implements EMP-417

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add disk_prepare test: grow rootfs overlapping userdata fails

Contributes to EMP-389, implements EMP-417

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add prepare_disk script test: grow boot with invalid start fail

Contributes to EMP-389 and EMP-406

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add prepare_disk script test: grow beyond disk end fail

Contributes to EMP-389 and EMP-406

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add prepare_disk script test: testing partition corruption restore

Contributes to EMP-389 and EMP-406

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add some more sanity check tests

- test script fails when partition table is not found
- test script fails when partition table is not found in the given system config dir
- test script fails is the given srorage device is not a block device
- test script fails when block device is not provided

Contributes to EMP-389 and EMP-406

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak commented:
Do not merge before other pull requests are part of master. 
This branch should remove the first commit 3f91991 and rebase on master first

oliv3r commented on .gitlab-ci.yml:
```diff
@@ -162,6 +163,7 @@ build:master_docker_latest:
     - build:master_docker_latest
   script:
     - ./test/toolbox_image.sh .build_*/um-update*.xz.img
+    - ./test/prepare_disk.sh .build_*/um-update*.xz.img
```
for our tests, i'd recommend putting each in its own job. feel free to template it so it only reads:
```yaml
test:toolboxtest_1/2:
<<: *job_template
script: ./test/toolbox_image.sh ...
```
or something. a) it allows for the tests to run in parallel; b) you more easily see which tests fails in the overview, rather then 'all tests failed'

raysiudak replied:
Yes we could do that, but.... this will result in having to add the tests in the yaml file for every build job type, i.e. 

test:dev_docker_candidate
test:dev_docker_latest
test:master_docker_latest

So both solutions are not ideal, lets park this for now in EMP-374 and pick this up when we setup all the repositories properly.

oliv3r commented on test/include/armhf_emulation.sh:
```diff
@@ -0,0 +1,99 @@
+#!/bin/sh
```
i think this was part of the 417 review so ignoring it here

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,431 @@
+#!/bin/sh
```
you reccon it a good idea to name the script scripts/30_prepare_disk here to improve searchability etc?

raysiudak replied:
No i don't think so, i think adding the number for sorting should only be added if this is intended and not for the reason of script names having the exactly tje same filename.

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,431 @@
+result=0
+
+# shellcheck source=test/include/armhf_emulation.sh
+. "test/include/armhf_emulation.sh"
```
put the include at the top of the file, ideally before the set -eu bit

raysiudak replied:
I moved this to the top, but it should be under the 'set' statement because you want the including script to control this behaviour and the include script introduces variables that potentially be unset.

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,431 @@
+. "test/include/armhf_emulation.sh"
+
+
+execute_prepare_disk()
```
i assume everything below is just a copy/paste from the other script, so ignoring that

raysiudak replied:
Yes except for some variable naming and the location of the temporary files.

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,431 @@
+
+test_disk_integrity()
+{
+    set -e
```
hmm, this looks like a change, and why is that needed on a per test basis? Either the set -eu (what happened to the u) is applied whenever the file is sourced, or from the sourcing file, i'd be supprised that you have to set it each call to the function. (Unless you are doing if test_function; then) then, i understand why you are doing this, but then we should really re-evaluate our testing framework.

raysiudak replied:
Yes this is wrong. The set statement is now only in the top of the file before everything else, all functions expected to exit on error are not executed in an 'if' statement. 
Tests however are executed within an 'if' statement and should return something

raysiudak committed:
Removed script name from test function

Since the test scripts is intended to run tests for only that script,
the script name is no longer needed as part of the test function.

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r commented on test/include/chroot_env.sh:
```diff
@@ -27,7 +27,7 @@ setup_chroot_env()
     # in the ro squashfs for testing purposes,
     # e.g. ARM_EMU_BIN mount point and temporary test files.
-    overlayfs_dir="$(mktemp -d -t "overlayfs.XXXXXXXXXX")"
+    overlayfs_dir="$(mktemp -d -t "overlayfs.XXXXXX")"
```
I think this was an unintended change whilst resolving.

raysiudak replied:
This is intended, six random positions are enough

oliv3r replied:
i'll mark it as resolved, sine this will be merged via EMP-407

oliv3r commented on .gitlab-ci.yml:
```diff
@@ -162,7 +162,8 @@ build:master_docker_latest:
   script:
-    - ./test/toolbox_image.sh ".build_"*"/um-update"*".xz.img"
+    - ./test/toolbox_image.sh .build_*/um-update*.xz.img
+    - ./test/prepare_disk.sh .build_*/um-update*.xz.img
```
I think this was an inintended change from merging, also also quote the new addition :)

raysiudak replied:
This is not intended.

oliv3r replied:
as this was also fixed in EMP-407, marking as resolved.

oliv3r commented on test/include/chroot_env.sh:
```diff
@@ -76,13 +76,13 @@ teardown_chroot_env()
         fi
-        if [ -d "${mount}" ] && [ -z "${mount##*overlayfs_*}" ]; then
-            rm -r "${mount:?}"
+        if [ -d "${mount}" ] && [ -z "${mount##*overlayfs*}" ]; then
```
p.s. you could also put your template/mask in a variable and reference it in the needed spots :p (just a tip, :tipping_hand_man:)

oliv3r replied:
ditto, per EMP-407

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,431 @@
+shift "$((OPTIND - 1))"
+
+if [ "${#}" -ne 1 ]; then
+    echo "Missing argument <file.img>."
```
Hmm, only spotted this by accident here, but did you not rename this before?

raysiudak replied:
I changed it in the usage, not in this place

oliv3r replied:
Not sure this was changed here, EMP-407 did change it elsewhere I think

oliv3r committed:
Make mktemp templates consistent again

With commit c6b9d44 ("Safe removal of temporary directories") we added
some templates for mktemp, but where not consistent herein. This patch
fixes that.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -0,0 +1,254 @@
+
+setup()
+{
+    toolbox_root_dir="$(mktemp -d -t "um-update-toolbox.XXXXXXX")"
```
why did you not use the name template that you used with the update_files script? would have been nice and consistent :p

oliv3r commented on .gitlab-ci.yml:
```diff
@@ -163,8 +163,8 @@ build:master_docker_latest:
     - ./test/toolbox_image.sh ".build_"*"/um-update"*".xz.img"
-    - ./test/update_files.sh ".build_"*"/um-update"*".xz.img"
     - ./test/prepare_disk.sh ".build_"*"/um-update"*".xz.img"
+    - ./test/update_files.sh ".build_"*"/um-update"*".xz.img"
```
since this is part of your first commit, you could have just fixed it there.

oliv3r commented on test/include/chroot_env.sh:
```diff
@@ -29,7 +29,7 @@ setup_chroot_env()
     # in the ro squashfs for testing purposes,
     # e.g. ARM_EMU_BIN mount point and temporary test files.
-    overlayfs_dir="$(mktemp -d -t "${NAME_TEMPLATE_OVERLAYFS}.XXXXXXXX")"
+    overlayfs_dir="$(mktemp -d -t "${NAME_TEMPLATE_OVERLAYFS}.XXXXXX")"
```
this is my bad, i fixed it in 407, but messed up on that eventually, sorry.

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -126,13 +126,13 @@ teardown()
     cd "${CWD}"
 
-    if [ -d "${work_dir}" ] && [ -z "${dir##*prepare_disk_work_dir_*}" ]; then
+    if [ -d "${work_dir}" ] && [ -z "${work_dir##*prepare_disk_work_dir*}" ]; then
```
this fix isn't part of a merge conflict or this commit!

oliv3r commented on test/prepare_disk.sh:
```diff
@@ -379,7 +379,7 @@ done
 
 if [ "${#}" -ne 1 ]; then
-    echo "Missing argument <file.img>."
+    echo "Missing argument <toolbox image file>."
```
sneaked this change unrelated to this changeset in here :) and completly unessasery as you added this in the first commit of this changeset.

oliv3r reviewed (commented):
The commit hygiene 72da14d548e402a9b ("Removed script name from test function, since this test scripts is 
intended to run tests for only that script.") is in need of some improvement.

raysiudak commented:
I am confused about what the state is of this pull request, what is still requried.
